src/cart/crt_group.c

```diff
@@ -316,19 +321,85 @@ grp_li_uri_set(struct crt_lookup_item *li, int tag, const char *uri)
 		ui->ui_ref = 0;
 		ui->ui_initialized = 1;
 		ui->ui_rank = li->li_rank;
-		D_STRNDUP(ui->ui_uri[tag], uri, CRT_ADDR_STR_MAX_LEN);
-		if (!ui->ui_uri[tag]) {
-			D_FREE_PTR(ui);
-			D_GOTO(exit, rc = -DER_NOMEM);
+
+		if (crt_provider_is_contig_ep(crt_gdata.cg_na_plugin)) {
+			D_STRNDUP(tmp_uri, uri, CRT_ADDR_STR_MAX_LEN);
+			if (!tmp_uri)
+				D_GOTO(exit, rc = -DER_NOMEM);
+
+			/* For now we assume contiguous endpoint providers are
+			 * port based. Based on that we generate URIs for every
+			 * tag of the rank from the base port.
+			 *
+			 * Port-based providers have form of
+			 * string:port
+			 *
+			 * Parse both parts out
+			 */
+			p = tmp_uri;
+			while (*p != '\0')
+				p++;
+			while (*p != ':' && p != tmp_uri)
+				p--;
+
+			if (p == tmp_uri) {
+				D_ERROR("Badly formed URI '%s'\n", tmp_uri);
+				D_FREE(tmp_uri);
+				D_GOTO(exit, rc = -DER_INVAL);
+			}
+
+			/* Split <string> from <port> part in URI */
+			*p = '\0';
+			p++;
+			base_port = atoi(p) - tag;
+
+			if (base_port <= 0) {
+				D_ERROR("Failed to parse uri=%s correctly\n",
+					tmp_uri);
+				D_FREE(tmp_uri);
+				D_GOTO(exit, rc = -DER_INVAL);
+			}
+
+			for (i = 0; i < 255; i++) {
+				char *tag_uri = NULL;
+
+				D_ASPRINTF(tag_uri, "%s:%d", tmp_uri,
+					   base_port + i);
+
+				if (tag_uri == NULL) {
+					int k;
+					for (k = 0; k < i; k++)
```

Comment:
(style)  Missing a blank line after declarations

Reply:
In crt_context, you used max_size -1 for STRNDUP,  here you are using max_size.  Why the difference?

Reply:
Why not just used strrchr(tmp_uri,":")

Reply:
If we know maximum possible size, then just use local array and avoid the dup.  Will make code faster avoiding the allocation and freeing of memory.

Reply:
yes should be size-1 here as well to avoid complains from coverity scans, will change.
regarding strrchr - yes could use it, but code scans tend to complain about majority of str* functions, but will change it and see if we get flagged; if so will change it later back to manual parsing.
and yes, using tmp_uri on a stack with fixed size is better, will change.

Reply:
Be care full of linkage of strndup.  The implementation in  build/external/dev/ofi/fabtests/include/windows/osd.h is incorrect.  The implementation in  build/external/dev/ofi/include/windows/osd.h is correct.

Reply:
We do not link either of those